### DISCUSSION TITLE
Explanation of type matching in `ref_test` tests

### DISCUSSION THREAD

**Discussion by q82419** [WebAssembly/gc#511] [Q&A] [answered]
Hi, when running the [`ref_test` test suite here](https://github.com/WebAssembly/gc/blob/main/test/core/gc/ref_test.wast#L182-L330), I have some questions.

When validating this module, we have defined types:

```
  (type $t0 (sub (struct)))
  (type $t1 (sub $t0 (struct (field i32))))
```

And in the `"test-sub"` function:

```
  (func (export "test-sub")
    (call $init)
    (block $l
      ;; must hold
      (br_if $l (i32.eqz (ref.test (ref null $t0) (ref.null struct))))
      (br_if $l (i32.eqz (ref.test (ref null $t0) (ref.null $t0))))
      (br_if $l (i32.eqz (ref.test (ref null $t0) (ref.null $t1))))   ;; mark here
      (br_if $l (i32.eqz (ref.test (ref null $t0) (ref.null $t2))))
      (br_if $l (i32.eqz (ref.test (ref null $t0) (ref.null $t3))))
     ...
```

Let's focus on the `ref.test` instruction in the marked line `(ref.test (ref null $t0) (ref.null $t1))`.

First the `ref.null $t1` will push the type `ref null $t1` onto the stack by the [validation rule](https://webassembly.github.io/gc/core/valid/instructions.html#xref-syntax-instructions-syntax-instr-ref-mathsf-ref-null-mathit-ht).

And according to the [rule of `ref.test`](https://webassembly.github.io/gc/core/valid/instructions.html#xref-syntax-instructions-syntax-instr-ref-mathsf-ref-test-mathit-rt), this will be valid for any valid reference type `rt'` on the top of stack for which `rt` matches `rt'`. 

That means, in this case, this instruction will be valid for which `ref null $t0` (of this instruction) matches `ref null $t1` (on the top of the stack).

That is, this instruction will be valid for which `(sub (struct))` matches `(sub $t0 (struct (field i32)))`.

Can you help to explain why this validation passes, or if I misunderstand the rules? Thanks!

**Comment by rossberg** [comment, marked as answer]
By [subsumption](https://webassembly.github.io/gc/core/valid/instructions.html#subsumption-for-xref-syntax-instructions-syntax-instr-mathit-instr-ast), the instruction `(ref.null $t1)` also has type `[] → [(ref null any)]`, and `(ref null any)` is a suitable rt' that is a supertype of `(ref null $t0)`.

This is a bit of a trick of the kind sometimes used in the specification of type systems. What the rule is saying is that you can weaken the operand type to anything less precise than rt (since it's a downcast anyway). Implicitly, what it is _really_ saying is that you can always pick the corresponding top type. Before we introduced the different subtype hierarchies (any/func/extern), `ref.test` in fact simply had type `[anyref] → [i32]`. The current formulation allows picking funcref or externref instead as the top type suitable for rt.

Since this is not so obvious, I created #513 that adds a note to this effect and a case to the validation algorithm in the Appendix.

**Reply by q82419** [reply]
Hi @rossberg , thanks for your answer.

A further question. After passing the validation and invoke the `"test-sub"` function, the instructions [this line (222)](https://github.com/WebAssembly/gc/blob/main/test/core/gc/ref_test.wast#L222):

```
(br_if $l (i32.eqz (ref.test (ref null $t1) (ref.null $t0))))
```

Focus on the `(ref.test (ref null $t1) (ref.null $t0))`.

First according to the [`ref.null`](https://webassembly.github.io/gc/core/exec/instructions.html#xref-syntax-instructions-syntax-instr-ref-mathsf-ref-null-x) instruction, push the value `ref null $t0` onto the stack.

Then in [`ref.test (ref null $t1)`](https://webassembly.github.io/gc/core/exec/instructions.html#xref-syntax-instructions-syntax-instr-ref-mathsf-ref-test-mathit-rt), pop that value (`ref null $t0`), and push `1` if the reference type of the value matches `ref null $t1`, `0` if not.

For passing this test, the `i32.eqz` should be `0`, while the `ref.test` should be `1`.

That is, the `ref null $t0` should matches `ref null $t1` for passing the test.

As mentioned above, the `ref null $t0` not matches `ref null $t1`.

Maybe as mentioned in [discussion 507](https://github.com/WebAssembly/gc/discussions/507), the `ref.null` instruction pushes the least type here. But as the spec, this case occurs only when the `ht` of `ref.null` is abstract heap type.

Can you help to answer this? Thanks!

**Reply by rossberg** [reply]
Yes, this succeeds per the discussion in #507.

It is important to distinguish static and dynamic checking. Static checking is only an approximation. At runtime, an operand may have a more precise type, i.e., a subtype of what's known statically. That implies that a cast can succeed even though the static types do not match. In particular, the dynamic type of ref.null is a subtype of its static type, for the stupid reasons mentioned in #507.

**Reply by rossberg** [reply]
> But as the spec, this case occurs only when the ht of ref.null is abstract heap type.

I'm not sure what you mean by that. The corresponding [dynamic typing rule for ref.null](https://webassembly.github.io/gc/core/exec/values.html#null-references-xref-syntax-instructions-syntax-instr-ref-mathsf-ref-null-t) applies for all ht.

**Reply by q82419** [reply]
Sorry that not specified so detail.

As the note in [`ref.null` instruction](https://webassembly.github.io/gc/core/exec/instructions.html#xref-syntax-instructions-syntax-instr-ref-mathsf-ref-null-x):

```
No formal reduction rule is required for the case `ref.null absheaptype`, since the instruction form is already a value.
```

I recognized that only for the abstract heap type case, the dynamic typing rule will be applied.
Since the all `ht` are applied the dynamic typing rule, why not just push `ref null t, t ∈ {none, nofunc, noextern}` in step 4 of execution?
Thanks.

**Reply by rossberg** [reply]
The note is just saying why we don't need a case for the formal(!) reduction rule where ht=absheaptype, because it would reduce to itself, as it already is a value.

The dynamic typing rule will be applied to all values, but `ref.null x` is not a value, so needs reducing first. This is mostly a technicality because it needs to rewrite its type annotation to be closed.

The reduction rule was a late addition, when the other parts already existed and were needed. But given that we now reduce (some) ref.null anyway, you are right that this could be factored somewhat differently. But it would be slightly more cumbersome to specify the syntax of values, and to pick the correct bottom type in such a modified reduction rule for ref.null would seemingly require a dynamic type check.

**Reply by q82419** [reply]
Thanks for the information.
Sorry that there're so many questions in GC spec, aspecially in the type matching.

**Reply by rossberg** [reply]
No worries, all good questions. This is much more subtle than everything we had in the type system before.
